### PR TITLE
Use Operator SDK framework

### PR DESCRIPTION
**Pull request by zroubalik**
Keda controller is rewritten to use operator-sdk framework. The controller logic was simplified removing informers, locks...

### Usage changes
- adding Operator SDK tooling https://github.com/kedacore/keda/issues/54 
- enable Keda to be deployed as namespaced/cluster-wide controller, ie. there can be deployed multiple Keda controllers, each serving specific namespace only

### Major internal changes
- changes on ScaledObject are handled by Reconcile loop,  it creates and manages lifecycle of the relevant HPA(in case of deploment) and then starting Scale Loop
- created dependent resources are properly deleted by Finalizer on ScaledObject
- operator is not holding any state in the code (previously `sync.RWMutex` in ScaleHandler, etc)
- External Metrics Names are stored in the ScaledObejct Status 
- not using Clientset anymore, but using [Controller Runtime Client API](https://github.com/operator-framework/operator-sdk/blob/master/doc/user/client.md)
- Metrics adapter was moved to the separate container, `GetExternalMetricNames()` and `GetScaledObjectMetrics()` are moved from ScaleHandler directly to the provider
- now using [Structured logging](https://www.client9.com/structured-logging-in-golang/) provided by [Operator SDK Logging](https://github.com/operator-framework/operator-sdk/blob/master/doc/user/logging.md), 
 - Debug log messages previously logged via `log.Debug(msg)` are now logged under _Level 1_, ie. `log.V(1).Info(msg)`, this allow us to add more granularity in logging in the future (introduce _Level 2/3/4_ if needed)
- Permissions were a little bit restricted, previous were to wide (cluster-admin for keda-operator SA)

### Deploying this version of Keda 
This implementation can be deployed by running these commands, it is expecting **existing** `keda` namespace 
```
git clone https://github.com/zroubalik/keda.git -b operator-sdk
cd keda
sed -i 's/kedacore/zroubalik/g' deploy/operator.yaml  
kubectl apply -f deploy/crds/keda.k8s.io_scaledobjects_crd.yaml
kubectl apply -f deploy/crds/keda.k8s.io_triggerauthentications_crd.yaml
kubectl apply -f deploy/
```

### TODO
* ~~set appropriate loging levels `V(?)` for log messages previously marked as ` Debug` sections in the code~~  (`Debug` -> `V(1)`)
* ~~tweak adapter logging~~
* ~~test adapter~~ (Kafka scaler working correctly)
* ~~test namespaced vs cluster wide operator~~

### CODE REVIEW

**Comment by zroubalik** [comment]
Into PR's description I've added instructions on how to deploy this code.

**Comment by zroubalik** [comment]
I've tweaked rest of the outstanding minor issues. I've tested this on OpenShift 4 using Kafka trigger, worked as expected. The code is ready for a Review and Merge. 

Helm chart update might be needed after merging this PR

**Comment by zroubalik** [comment]
I've tweaked rest of the outstanding minor issues. I've tested this on OpenShift 4 using Kafka trigger, worked as expected. The code is ready for a Review and Merge.

Helm chart update might be needed after merging this PR

**Comment by zroubalik** [comment]
@tomkerkhove you might be interested to take a look on this PR

**Comment by tomkerkhove** [comment]
Thanks for the ping @zroubalik but I'll leave it up to @ahmelsayed as I'm not that knowledgeable on this level.

**Comment by jeffhollan** [comment]
I'll spend some time this week to test out this fork, @ahmelsayed plans to review sometime in the next week

**Comment by jeffhollan** [comment]
Validated that this works with my demos around Service Bus and Rabbit MQ.  Thanks @zroubalik

File: build/Dockerfile.adapter [hunk 1]
```diff
@@ -0,0 +1,19 @@
+FROM registry.access.redhat.com/ubi8/ubi-minimal:latest
+
+ENV OPERATOR=/usr/local/bin/keda-adapter \
+    USER_UID=1001 \
+    USER_NAME=keda
+
+COPY build/_output/bin/keda-adapter ${OPERATOR} 
+
+COPY build/bin /usr/local/bin
```
Comment (ahmelsayed) [inline build/Dockerfile.adapter]: I think this line is not needed, right? It fails for me on both images on a `make build`. Removing it from both resolves the issue.

Comment (ahmelsayed) [reply build/Dockerfile.adapter]: Actually I think that folder is just missing the `user_setup` script.

Comment (zroubalik) [reply build/Dockerfile.adapter]: Good catch, thanks. `.gitignore` was too strict and prevented pushing those files. There's so much files, so I haven't noticed that.  I'll fix it in a few.

Comment (zroubalik) [reply build/Dockerfile.adapter]: @ahmelsayed should be fixed now, PTAL

**Comment by zroubalik** [comment]
@ahmelsayed @jeffhollan I have added one last minor change. The original cluster permissions were to wide (granting cluster-admin to the keda SA). I have restricted the permissions a little bit and have tested it and working for me. I might want to test it with your scenarios or at least one :)

File: go.mod [hunk 2]
```diff
@@ -1,115 +1,74 @@
 module github.com/kedacore/keda
 
-go 1.13
+go 1.13.1
```
Comment (joskfg) [inline go.mod:3]: The go version is supposed to be "go <MAJOR>.<MINOR>" this is producing that I'm not able to use this package as dependency for a controller.

Comment (zroubalik) [reply go.mod:3]: @joskfg https://github.com/kedacore/keda/pull/718

Comment (joskfg) [reply go.mod:3]: Thanks a lot!

**Pull request merged**